program/src/processor/mod.rs

```diff
@@ -144,3 +162,75 @@ macro_rules! require {
         require!( $constraint, $error, format!($message, $($args)+) );
     };
 }
+
+/// Unpacks an initialized account from the given data and
+/// returns a mutable reference to it.
+#[inline]
+pub fn unpack_initialized_mut<T: Pod + IsInitialized>(
+    data: &mut [u8],
+) -> Result<&mut T, ProgramError> {
+    let account = bytemuck::try_from_bytes_mut::<T>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(account.is_initialized(), ProgramError::UninitializedAccount);
+
+    Ok(account)
+}
+
+/// Unpacks the delegation information from either a `SolStakerStake` and `ValidatorStake`
+/// accounts.
+///
+/// This function will validate that the account data is initialized and deriavation matches
+/// the expected PDA derivation.
+#[inline]
+pub fn unpack_delegation_mut<'a, 'b, 'c, 'd>(
+    stake_data: &'a mut [u8],
+    stake: &'b Pubkey,
+    config: &'c Pubkey,
+    program_id: &'d Pubkey,
+) -> Result<&'a mut Delegation, ProgramError> {
+    let (delegation, derivation) = match &stake_data[..ArrayDiscriminator::LENGTH] {
+        SolStakerStake::SPL_DISCRIMINATOR_SLICE => {
+            let sol_staker = unpack_initialized_mut::<SolStakerStake>(stake_data)?;
+
+            let (derivation, _) =
+                find_sol_staker_stake_pda(&sol_staker.stake_state, config, program_id);
+
+            (&mut sol_staker.delegation, derivation)
+        }
+        ValidatorStake::SPL_DISCRIMINATOR_SLICE => {
+            let validator = unpack_initialized_mut::<ValidatorStake>(stake_data)?;
+
+            let (derivation, _) =
+                find_validator_stake_pda(&validator.delegation.validator_vote, config, program_id);
+
+            (&mut validator.delegation, derivation)
+        }
+        _ => return Err(ProgramError::InvalidAccountData),
+    };
+
+    require!(stake == &derivation, ProgramError::InvalidSeeds, "stake");
+
+    Ok(delegation)
+}
+
+/// Unpacks the delegation information from either a `SolStakerStake` and `ValidatorStake`
+/// accounts.
+///
+/// This function will validate that the account data is initialized.
+#[inline]
+pub fn unpack_delegation_mut_uncheked(stake_data: &mut [u8]) -> Result<Delegation, ProgramError> {
```

Comment:
My other PR didn't change this to ref, can this be
```suggestion
pub fn unpack_delegation_mut_uncheked(stake_data: &mut [u8]) -> Result<&mut Delegation, ProgramError> {
```

Reply:
I removed some of the lifetimes that clippy said they could be inferred.